rcgen/src/key_pair.rs

```diff
@@ -319,6 +325,138 @@ impl KeyPair {
+		let private_key = pem::parse(pem_str)._err()?;
+		let private_key: &[_] = private_key.contents();
+		Self::from_der_and_sign_algo(
+			&PrivateKeyDer::try_from(private_key).map_err(|_| Error::CouldNotParseKeyPair)?,
+			alg,
+		)
+	}
+
+	/// Obtains the key pair from a DER formatted key
+	/// using the specified [`SignatureAlgorithm`]
+	///
+	/// If you have a [`PrivateKeyDer`], you can usually rely on the [`TryFrom`] implementation
+	/// to obtain a [`KeyPair`] -- it will determine the correct [`SignatureAlgorithm`] for you.
+	/// However, sometimes multiple signature algorithms fit for the same DER key. In those instances,
+	/// you can use this function to precisely specify the `SignatureAlgorithm`.
+	///
+	/// You can use [`rustls_pemfile::private_key`] to get the `key` input. If
+	/// you have already a byte slice, just calling `try_into()` will convert it to a [`PrivateKeyDer`].
+	///
+	/// [`rustls_pemfile::private_key`]: https://docs.rs/rustls-pemfile/latest/rustls_pemfile/fn.private_key.html
+	#[cfg(all(feature = "crypto", feature = "aws_lc_rs"))]
+	pub fn from_der_and_sign_algo(
+		key: &PrivateKeyDer<'_>,
+		alg: &'static SignatureAlgorithm,
+	) -> Result<Self, Error> {
+		let is_pkcs8 = matches!(key, PrivateKeyDer::Pkcs8(_));
+
+		let rsa_key_pair_from = if is_pkcs8 {
+			RsaKeyPair::from_pkcs8
+		} else {
+			RsaKeyPair::from_der
+		};
+
+		let serialized_der = key.secret_der().to_vec();
+
+		let kind = if alg == &PKCS_ED25519 {
+			KeyPairKind::Ed(Ed25519KeyPair::from_pkcs8_maybe_unchecked(&serialized_der)._err()?)
+		} else if alg == &PKCS_ECDSA_P256_SHA256 {
+			KeyPairKind::Ec(ecdsa_from_private_key_der(
+				&signature::ECDSA_P256_SHA256_ASN1_SIGNING,
+				&serialized_der,
+			)?)
+		} else if alg == &PKCS_ECDSA_P384_SHA384 {
+			KeyPairKind::Ec(ecdsa_from_private_key_der(
+				&signature::ECDSA_P384_SHA384_ASN1_SIGNING,
+				&serialized_der,
+			)?)
+		} else if alg == &PKCS_ECDSA_P521_SHA512 {
+			KeyPairKind::Ec(ecdsa_from_private_key_der(
+				&signature::ECDSA_P521_SHA512_ASN1_SIGNING,
+				&serialized_der,
+			)?)
+		} else if alg == &PKCS_RSA_SHA256 {
+			let rsakp = rsa_key_pair_from(&serialized_der)._err()?;
+			KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA256)
+		} else if alg == &PKCS_RSA_SHA384 {
+			let rsakp = rsa_key_pair_from(&serialized_der)._err()?;
+			KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA384)
+		} else if alg == &PKCS_RSA_SHA512 {
+			let rsakp = rsa_key_pair_from(&serialized_der)._err()?;
+			KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA512)
+		} else if alg == &PKCS_RSA_PSS_SHA256 {
+			let rsakp = rsa_key_pair_from(&serialized_der)._err()?;
+			KeyPairKind::Rsa(rsakp, &signature::RSA_PSS_SHA256)
+		} else {
+			panic!("Unknown SignatureAlgorithm specified!");
+		};
+
+		Ok(KeyPair {
+			kind,
+			alg,
+			serialized_der,
+		})
+	}
+
+	/// Parses the key pair from the DER format
+	///
+	/// You can use [`rustls_pemfile::private_key`] to get the `key` input. If
```

Comment:
I don't think we need rustdoc for this internal function outside of the first line.